language/move-prover/bytecode/src/access_path.rs

```diff
@@ -659,40 +659,29 @@ impl AccessPath {
-                // offsets of the current access path until a result is found
-                // or we do not have any more offsets to add
-                let addrs = loop {
-                    let ap = AccessPath::new(caller_root.clone(), offsets.clone());
-
-                    if let Some(addr) = sub_map.get_access_path(ap) {
-                        break addr;
-                    }
-
-                    // If we did not find an access path for the given formal
-                    // we return an empty set
-                    if new_offsets.is_empty() {
-                        return AbsAddr::default();
+                // offsets of the current access path.
+                // When a result is found, we accumulate to results
+                // We need to try with the offsets to handle cases such as
+                // Formal(1): { data: None, children: { Field(0): { data: Footprint(AccessPath { root: Formal(0), offsets: [] }) } }
+                // where simply reading Formal(1) would yield None
+                // but Formal(1)/0 gives the mapping to Formal(0)
+                for i in 0..=new_offsets.len() {
+                    let caller_offsets = Vec::from(&new_offsets[0..i]);
+                    let ap = AccessPath::new(caller_root.clone(), caller_offsets.clone());
+                    if let Some(addrs) = sub_map.get_access_path(ap) {
+                        // We need to adjust the offset from the current access path
+                        // to avoid duplicating it (i.e. Formal(0)/0/0 instead of Formal(0)/0)
+                        let callee_offsets = Vec::from(&new_offsets[i..new_offsets.len()]);
+                        let new_addrs =
+                            AccessPath::new(self.root.clone(), callee_offsets).prepend_addrs(addrs);
+                        results.join(&new_addrs);
                     }
-
-                    // If we did not find the data, try with the next offset
-                    // this is to handle cases such as
-                    // Formal(1): { data: None, children: { Field(0): { data: Footprint(AccessPath { root: Formal(0), offsets: [] }) } }
-                    // where simply reading Formal(1) would yield None
-                    // but Formal(1)/0 gives the mapping to Formal(0)
-                    // we need to remove the offset from the current access path
-                    // to avoid duplicating it (i.e. Formal(0)/0/0 instead of Formal(0)/0)
-                    // remove(0) is inefficient but should not be problematic given offsets is very small
-                    offsets.push(new_offsets.remove(0));
-                };
-
-                // new_offsets will either be the full offsets of this access path
-                // or a subset of them if we had to use some of them to read
-                // the access path
-                AccessPath::new(self.root.clone(), new_offsets).prepend_addrs(addrs)
+                }
+                results
```

Comment:
@sblackshear Hopefully, this should cover the edge case we just discussed! I'll add a testcase for that

Reply:
@sblackshear I'm struggling a bit to find a test that would cover this case nicely. Let me know if you have something on the top of your head

Reply:
I think something like
```
struct S { f: address, g: address }
struct A { a: address }

fun two_secondary_indexes(s: &S, b: bool): address {
  if (b) {
    borrow_global<A>(s.f).a
  } else {
    borrow_global<A>(s.g).a
  }
}
```

will exercise this (but not positive)

Reply:
I added the test but looking at the output I think it's not triggering the condition we discussed, the result looks correct with the previous implementation that exited the loop. I'll try to play a bit around with this.

Reply:
Ah, interesting... One more try;

```
struct A { b : B }
struct B { addr: address }

fun write_both_fields(a: &A, b: B, addr: address, flag: bool) {
  if (flag) {
    a.b = b
  }
  if (flag) {
    a.b.addr = addr
  }
}
```

Reply:
I tried it but it also gives the same result with both implementations. I'm yet to find something that fails with the previous implementation but I am pretty sure there are cases where it does happen :thinking: 
I added the test to the test suite nevertheless